spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala

```diff
@@ -0,0 +1,142 @@
+
+
+class CustomCatalogSuite extends QueryTest with SharedSparkSession
+  with DeltaSQLCommandTest {
+
+  override def sparkConf: SparkConf =
+    super.sparkConf.set("spark.sql.catalog.dummy", classOf[DummyCatalog].getName)
+
+  test("RESTORE a table from DummyCatalog") {
+    val tableName = "table1"
+    withTable(tableName) {
+      sql("SET CATALOG dummy")
+      val dummyCatalog =
+        spark.sessionState.catalogManager.catalog("dummy").asInstanceOf[DummyCatalog]
+      val tablePath = dummyCatalog.getTablePath(tableName)
+      sql(f"CREATE TABLE $tableName (id bigint) USING delta")
+      // Insert some data into the table in the dummy catalog.
+      // To make it simple, here we insert data directly into the table path.
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (0)")
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (1)")
+      sql(f"RESTORE TABLE $tableName VERSION AS OF 0")
+      checkAnswer(spark.table(tableName), Nil)
+      sql(f"RESTORE TABLE $tableName VERSION AS OF 1")
+      checkAnswer(spark.table(tableName), spark.range(1).toDF())
+      // Test 3-part identifier
+      sql(f"RESTORE TABLE dummy.default.$tableName VERSION AS OF 2")
+      checkAnswer(spark.table(tableName), spark.range(2).toDF())
+
+      sql("SET CATALOG spark_catalog")
+      // Test 3-part identifier when the current catalog is the default catalog
+      sql(f"RESTORE TABLE dummy.default.$tableName VERSION AS OF 1")
+      checkAnswer(spark.table(f"dummy.default.$tableName"), spark.range(1).toDF())
+    }
+  }
+}
+
+class DummyCatalog extends TableCatalog {
+  private val spark: SparkSession = SparkSession.active
+  private val tempDir: Path = new Path(Utils.createTempDir().getAbsolutePath)
+  // scalastyle:off deltahadoopconfiguration
+  private val fs: FileSystem =
+    tempDir.getFileSystem(spark.sessionState.newHadoopConf())
+  // scalastyle:on deltahadoopconfiguration
+
+  override def name: String = "dummy"
+
+  def getTablePath(tableName: String): Path = {
+    new Path(tempDir, tableName)
+  }
+  override def defaultNamespace(): Array[String] = Array("default")
+
+  override def listTables(namespace: Array[String]): Array[Identifier] = {
+    val status = fs.listStatus(tempDir)
+    status.filter(_.isDirectory).map { dir =>
+      Identifier.of(namespace, dir.getPath.getName)
+    }
+  }
+
+  override def tableExists(ident: Identifier): Boolean = {
+    val tablePath = getTablePath(ident.name())
+    fs.exists(tablePath)
+  }
+  override def loadTable(ident: Identifier): Table = {
+    val tablePath = getTablePath(ident.name())
+    DeltaTableV2(spark, tablePath)
+  }
+
+  override def createTable(
+    ident: Identifier,
+    schema: StructType,
+    partitions: Array[Transform],
+    properties: java.util.Map[String, String]): Table = {
+    val tablePath = getTablePath(ident.name())
+    // Create an empty Delta table on the tablePath
+    spark.range(0).write.format("delta").save(tablePath.toString)
```

Comment:
Shouldn't this use the passed-in schema? Otherwise it takes a DDL later to update it with info we already had...

Reply:
Alternatively, we can just create the DTV2 and call it good -- Delta knows how to handle the new/empty/missing directory case.

Reply:
> Shouldn't this use the passed-in schema?

Since it is a dummy catalog, I try to make the schema fixed as "id: long"

> Alternatively, we can just create the DTV2 

Could you show me some details of how to use it?

Reply:
`DeltaTableV2(spark, tablePath.toString)` should suffice? 
See [DeltaTableV2.scala](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala#L53)